improve 493: add  support in odata (Draft)

sadiqkhoja opened the pull request:
Added `$select` support in odata endpoint.

matthew-white commented on lib/formats/odata.js:
```diff
@@ -418,6 +430,48 @@ const singleRowToOData = (fields, row, domain, originalUrl, query) => {
+        // it's ok to include field with repeat type so that user can have navigation link
+        // but child field of a repeat field is not supported
+        if (field.type === 'repeat' && i < propSegments.length - 1) throw Problem.user.unsupportedODataSelectField({ field: property });
+        filteredFields.add(field);
```
I'm not totally sure, but I think that if the user `$select`s a field in a group, but doesn't `$select` the group, we should also add the group to `filteredFields`. For example: `$select=meta/instanceID` instead of `$select=meta,meta/instanceID`. Otherwise I think `generateDateFrame()` won't work as expected.

sadiqkhoja replied:
This is handled by adding group fields automatically to `filteredFields`

sadiqkhoja replied:
Tested in `should return a single row result with selected properties` integration test

matthew-white commented on lib/formats/odata.js:
```diff
@@ -418,6 +430,48 @@ const singleRowToOData = (fields, row, domain, originalUrl, query) => {
+// Validates $select query parameter including metadata properties and returns list of FormFields
+const filterFields = (formFields, select, table) => {
+  const filteredFields = new Set();
+  const fieldMap = formFields.reduce((map, field) => ({ ...map, [field.path.toLowerCase()]: field }), {});
```
The OData name of a field can be different from the name of the field in the form: see `sanitizeOdataIdentifier()`. I'm thinking that `fieldMap` should use those OData names to map.

sadiqkhoja replied:
Fixed it by sanitizing the path

matthew-white commented on lib/data/odata.js:
```diff
@@ -129,30 +130,45 @@ const submissionToOData = (fields, table, submission, options = {}) => new Promi
+    };
+
+    if (options.metadata) {
+      const systemProp = options.metadata.filter(p => p.startsWith('__system')).map(p => p.substr(9));
+      if (systemProp.length > 0) {
+        root.__system = {};
+        for (const property of systemProp) {
+          root.__system[systemObj[property].key] = systemObj[property].value;
+        }
```
I think the user could specify `$select=__system` to get all the data in `__system`: in OData, `__system` is itself a property that is a `ComplexType`.

sadiqkhoja replied:
Added this feature

sadiqkhoja replied:
will add the behavior for other complexType in a separate PR. It's quite complex

sadiqkhoja replied:
added this in https://github.com/getodk/central-backend/pull/717

matthew-white reviewed (commented):
Thanks, @sadiqkhoja! I left a few comments below, but this is definitely looking like it's going in the right direction. (By the way, sorry for getting these comments to you late: I was busy during the day, so decided to flip things around and work in the evening.)

I think your approach of implementing `$select` for `__id` and `__system` one way (using `options.metadata`) and form fields another way (using `selectFields()`) makes a lot of sense. I like your approach of filtering the fields within the endpoint, before passing the fields to `submissionToOData()`. I was thinking that `submissionToOData()` was going to have to have more logic around which form fields were `$select`ed (I thought `generateDataFrame()` might have to change, for example). But it'd be great if that's not needed! (`submissionToOData()` is already pretty complex without that logic.) I'm pretty sure that `submissionToOData()` can handle data in the submission XML that doesn't correspond to one of the `fields` passed to `submissionToOData()`.

It looks like you've implemented `$select` so that it's case-insensitive (and I can see that that's making things more complicated in a couple of places). Is it part of the OData spec that `$select` is case-insensitive? (If so, I wonder if that's the case for `$filter` as well. At the moment, our implementation of `$filter` is case-sensitive.)

sadiqkhoja commented:
Thanks @matthew-white I'll incorporate your comments soon

issa-tseng commented on lib/formats/odata.js:
```diff
@@ -70,8 +70,14 @@ const nextUrlFor = (limit, offset, count, originalUrl) =>
+const extractOptions = (query) => ({
+  wkt: isTrue(query.$wkt),
+  expand: query.$expand,
+  metadata: query.$select?.split(',')
+    .map(p => p.trim().toLowerCase())
+    .filter(p => p.startsWith('__'))
+});
```
it seems to me most of your use of the `metadata` object is as a lookup, not as a list. can you use a `{}` instead of an array here?

sadiqkhoja replied:
Thanks, I have refactored the code

sadiqkhoja commented on lib/formats/odata.js:
```diff
@@ -418,6 +430,48 @@ const singleRowToOData = (fields, row, domain, originalUrl, query) => {
+  '__system/deviceid',
+  '__system/edits',
+  '__system/formversion',
+]);
```
done

sadiqkhoja commented on lib/formats/odata.js:
```diff
@@ -418,6 +430,48 @@ const singleRowToOData = (fields, row, domain, originalUrl, query) => {
+  return Array.from(filteredFields);
+};
+
+const selectFields = (query, table) => (fields) => (query.$select ? resolve(filterFields(fields, query.$select, table)) : resolve(fields));
```
done

sadiqkhoja commented on lib/formats/odata.js:
```diff
@@ -418,6 +430,48 @@ const singleRowToOData = (fields, row, domain, originalUrl, query) => {
+      for (let i = 0; i < propSegments.length; i+=1) {
+        path += `/${propSegments[i]}`;
+        const field = fieldMap[path];
+        if (!field) throw Problem.user.propertyNotFound({ field: property });
```
Thanks for highlighting this. I have added code to handle property selection for subtables

sadiqkhoja commented on lib/formats/odata.js:
```diff
@@ -418,6 +430,48 @@ const singleRowToOData = (fields, row, domain, originalUrl, query) => {
+        if (!field) throw Problem.user.propertyNotFound({ field: property });
+        // it's ok to include field with repeat type so that user can have navigation link
+        // but child field of a repeat field is not supported
+        if (field.type === 'repeat' && i < propSegments.length - 1) throw Problem.user.unsupportedODataSelectField({ field: property });
```
Yes, you are right if $expand is used with $select then empty repeat groups are returned. 
Since only `$expand=*` is allowed, I can throw unsupported feature problem if user provides both `$select` and `$expand` or we can simply ignore `$select` option if `$expand` is also present, what do you recommend?

sadiqkhoja replied:
Added validation in `endpoint.js`

sadiqkhoja commented:
@matthew-white I have made following changes:

- corrected the handling of $select for subtable
- made $select case-sensitive
- added ability to provide just __system in $select
- added integration tests
- incorporated other feedback

Regarding case-sensitivity of query parameters like `$select', `$filter`, etc, the latest version (4.01) of OData specs require them to be case insensitive. In fact `$` sign is optional too. However, I couldn't find any requirement related to property names. In any case, I have removed `toLowerCase` for simplicity's sake. Maybe we can create separate issue to make OData options case-insensitivity across the board.

matthew-white reviewed (commented):
Thanks, @sadiqkhoja! This is looking good. The tests provide good coverage and look well organized.

You may have noticed this already, but it looks like one test is failing on CircleCI.

It's exciting that there's now support for `$select`ing properties in repeat group tables; I didn't think we'd be able to do that yet. I also like that we support `$select=*`. I think it was helpful for code clarity to make `$select` case-sensitive: there's already plenty going on in the code.

I also wanted to let you know that I noticed a separate issue while thinking about the code here: #496. I don't think that issue will affect your work here, but I thought I'd point it out.

sadiqkhoja commented on lib/formats/odata.js:
```diff
@@ -418,6 +430,44 @@ const singleRowToOData = (fields, row, domain, originalUrl, query) => {
+// Validates $select query parameter including metadata properties and returns list of FormFields
+const filterFields = (formFields, select, table) => {
+  const filteredFields = new Set();
+  const fieldMap = formFields.reduce((map, field) => ({ ...map, [`${field.path.replace(/[^/]*$/, '')}${sanitizeOdataIdentifier(field.name)}`]: field }), {});
```
done

sadiqkhoja commented on lib/formats/odata.js:
```diff
@@ -418,6 +430,44 @@ const singleRowToOData = (fields, row, domain, originalUrl, query) => {
+  for (const property of select.split(',').map(p => p.trim())) {
+    // validate metadata properties. __system/.. properties are only valid for Submission table
+    if (property.startsWith('__')) {
+      if (!(property === '__id' || (table !== 'Submissions' && property === `__${table.replace(/^(.*)\.[^.]+\.[^.]+$/, '$1').replace(/\./g, '-')}-id`) || (table === 'Submissions' && systemFields.has(property))))
```
I've simplified the condition. If user requests `__id` then I return all the IDs (parent and child)

matthew-white replied:
Unless the OData spec calls for that behavior, I think my preference would be just to return `__id`. I can see how it'd be nice to provide the user a way to access all IDs, but we're already providing a lot of useful `$select` functionality with this PR. I think we could come back to other IDs in a future PR.

sadiqkhoja replied:
done

sadiqkhoja commented on lib/formats/odata.js:
```diff
@@ -418,6 +430,44 @@ const singleRowToOData = (fields, row, domain, originalUrl, query) => {
+      if (table !== 'Submissions') {
+        for (const tableSegment of table.replace(/Submissions\./, '').split('.')) {
+          path += `/${tableSegment}`;
+          filteredFields.add(fieldMap[path]); // fieldMap is guaranteed to have structure field
```
fixed

matthew-white commented:
@sadiqkhoja, I'm still taking a look at your changes (I should be done before we meet tomorrow). In the meantime, I'm noticing that there are a few small conflicts. Would you mind taking a look at those? Also, now that #545 is merged, it'd be awesome if you could verify that linting passes for the new tests.

matthew-white commented on lib/formats/odata.js:
```diff
@@ -464,7 +475,16 @@ const filterFields = (formFields, select, table) => {
-  return Array.from(filteredFields);
+
+  const filteredFieldsArr = Array.from(filteredFields);
+  // Sort the fields by `order` as Set() doesn't gurantee an order
```
I think `Set`s and `Map`s iterate according to insertion order, in case that helps!

sadiqkhoja replied:
insertion order is based on the value of `$select`, which is defined by the user. So I have to order it here

matthew-white commented on lib/formats/odata.js:
```diff
@@ -350,10 +350,17 @@ const rowStreamToOData = (fields, table, domain, originalUrl, query, inStream, t
         for (const field of data) {
+
+          // if ID (of subtable) is not requested then remove it
+          const fieldRefined = (options.metadata && !options.metadata.__id) ? (({ [parentIdProperty]: __, ...item }) => item)(field) : field;
```
There's a utility function in lib/util/util.js named `without()` that I think would be helpful here.

matthew-white commented on lib/formats/odata.js:
```diff
@@ -344,10 +350,17 @@ const rowStreamToOData = (fields, table, domain, originalUrl, query, inStream, t
       // text resultstream, and call done to indicate that the row is processed.
       submissionToOData(fields, table, row, options).then((data) => {
+
+        const parentIdProperty = Object.keys(data[0] ?? true).filter(p => /^__.*-id$/.test(p))[0];
```
I had no idea that `Object.keys()` could be called with `true`! Is `true` automatically converted to an object? I feel like a ternary might be a little clearer here, something like:

```js
const parentIdProperty = data.length !== 0
  ? Object.keys(data[0]).find(p => /^__.*-id$/.test(p))
  : null;
```

(Also using `find()`, which I think is a little more performant than `.filter()[0]`.)

matthew-white commented on lib/formats/odata.js:
```diff
@@ -464,7 +475,16 @@ const filterFields = (formFields, select, table) => {
+  const filteredFieldsArr = Array.from(filteredFields);
+  // Sort the fields by `order` as Set() doesn't gurantee an order
+  filteredFieldsArr.sort((a, b) => a.order - b.order);
+  // Resequence the order because downstream function(s) are dependent on it e.g schema.children
```
Ah very interesting, good catch!

matthew-white commented on lib/formats/odata.js:
```diff
@@ -70,8 +69,15 @@ const nextUrlFor = (limit, offset, count, originalUrl) =>
+  metadata: query.$select !== '*' && query.$select?.split(',')
+    .map(p => p.trim())
+    .filter(p => p.startsWith('__'))
+    .reduce((map, field) => ({ ...map, [field]: true }), {})
```
I think it'd be a little better to initialize this as `Object.create(null)` instead of `{}`, since the keys are user-provided. Someone might have the bright idea of specifying `$select=__proto__`, in which case reasoning about this starts getting confusing. 😆 Or I think a `Set` would work well here!

matthew-white reviewed (commented):
These changes are looking good! I've added a few questions/comments below. Could you also document our support of `$select` in the API docs?

matthew-white reviewed (approved):
Looks great! ✨ I left a couple of very minor comments. Feel free to merge after taking a look at those!
